[DVDVideoCodecIMX] - New implementation

mk01 opened the pull request:
to provide partial release of framebuffers when media changed
parameters during play / decoder changes minBuffer demand.

without this - commit ea098fef388561e0f508ca691e7a381aaabea948 can lead to VPU mem drain in case decoder would trigger reconfigure more times during one session

fritsch commented:
Here I need to read a bit longer to clearly understand what it does. Thanks much in advance.

fritsch commented:
Now needs a rebase after the other candidates went in, please.

mk01 commented:
@fritsch 

the lock are going completely out (from the codec part) , just wait. just I'm not pushing the "pool" as wolfgar complained about more commit in one PR - but with more PRs I have mess here, ... so once merged, I will pull and pick next small part to PR.

this PR is nothing more than repositioned exiting code 

A way would be also to just rebase it onto new libimxvpu right now, then whole decoding part would shrink to fraction of the code - but this would cause just more stress to you again. 

so confirm this one, then it goes only better.

mk01 commented:
@fritsch 
rebased

fritsch commented:
It's not stress to me at all. You are the one that is actively working on it, so it's your say at the end ...

Btw. if you want to rework the whole decoder, perhaps a complete "reimplementation" PR would be better. It's perfectly fine for me to add several commits. I only want to see the big picture of where we are going to.

mk01 commented:
@fritsch 
ok, then I will add the others

MartijnKaijser commented:
go for merge?

fritsch commented:
@mk01 please rebase (we changed some ffmpeg defnines) and can you summarize the current work?

mk01 commented:
@FernetMenta 
after the refclock rebrand and removing Wait() you still kept SendVblankSignal() - do you have any specific plans with it?

Basically I need the same as @popcornmix - blocking vsync call to place into HwRenderer/RendererIMX.

FernetMenta commented:
@mk01 SendVblankSignal() is a left over. I don't have plans with this.

mk01 commented:
@FernetMenta 
would be great if you can share your idea how to deal with the need to slow down the render loop along the way so I can go the right direction. for now I implemented this:
https://github.com/xbmc/xbmc/pull/9253/commits/33e65eb4f087253eb39d7a26bcee7373b426fd78 ?

FernetMenta commented:
@mk01 the strategy is rather simple: the render loop is supposed to cycle one per vblank interval. If you bypass gui rendering, you renderer has to take care for this.
Abusing videoRefClock for this purpose is wrong. There is no reason for renderer to know about VideoRefClock object. I suggest you do something similar as Pi.

mk01 commented:
@FernetMenta 
so your proposal is to create completely new entity with copy of code refclock object is already providing, even more to double executed code just to provide something refclock base is already running anyhow?

FernetMenta commented:
No this is absolutely not was I am suggesting. Now IMX has an additional client for the vblank logic. Why don't you factor this code out and use it in locations that require it. Like Pi. That's much cleaner than hacking VideoRefClock.

mk01 committed:
[RenderIMX] fix 100% load if display fps != media fps

mk01 committed:
[CodecIMX] refactor to sync Blit, tripple buffer, render via IPU, use tiled format,
           add cropping intermediate IPU task

mk01 committed:
[CodecIMX] move common IMX buffer definition to imx's global defs - IMX.h

mk01 committed:
[CodecIMX] update de-interlacing modes

mk01 committed:
[IMX] template for blocking FIFO buffer

mk01 committed:
[CodecIMX]

- refactored decoder into separate thread
- simplified the workflow around decoder and output
- removed all locking around frames (only reference count is used)
- added input (from demuxer) and output (to player) buffers
- changed format of VPU output to frame tile

 (some other details:
  - removed counting of consumed bytes, no longer tracking
  framebuffers, decoder can reconfigure its params/memory
  onthefly no without restart/or reseting input/output.
  - updated to use std::atomic ...)

mk01 committed:
[IMX] fix Capture && Screenshot

mk01 committed:
[CodecIMX] padding to avoid IPU-IC do split -> 544 + 2x upscale when cropping 1088lines.

fritsch commented:
@mk01 short ping here. We are heading towards beta. Can you give us a status update here? Also another question: Was IMX fully fixed in regard to the RenderUpdate rework now running with display refreshrate?

I would like to get this rework into v17 if possible.

mk01 committed:
[IMX] put time limit on VSYNC wait

mk01 committed:
[IMX] change flow on EGL/Codec initialisation

mk01 committed:
[IMX] fix bug when interlaced content was started as 'minimized' (eg LiveTV)

mk01 committed:
[IMX] additional tuning / fixups

mk01 commented:
@fritsch 

yes, the specific case when refresh rate switching is off so main renderer runs at display's rate is no issue anymore.

fritsch commented:
@mk01: Reading the code I missed the blocking for WaitVsync() after render,
did I overlook that?

2016-07-19 3:50 GMT+02:00 Matus Kral <notifications@github.com>:

> @fritsch <https://github.com/fritsch>
>
> yes, the specific case when refresh rate switching is off so main renderer
> runs at display's rate is no issue anymore.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/9253#issuecomment-233509101>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHVAk4brusGZJgt5OSfJrtd5PkJikks5qXC1RgaJpZM4HoAcC>
> .
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

mk01 commented:
@fritsch 
https://github.com/xbmc/xbmc/pull/9253/files#diff-857d15e9c9c7903fce4f77400fe63cfaR164

fritsch commented:
Perfect :-) thanks much.

So anything you want concerning IMX in v17? If yes, let us know.

2016-07-19 16:24 GMT+02:00 Matus Kral <notifications@github.com>:

> @fritsch <https://github.com/fritsch>
>
> https://github.com/xbmc/xbmc/pull/9253/files#diff-857d15e9c9c7903fce4f77400fe63cfaR164
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/9253#issuecomment-233648884>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHXqRPS3HosSIVm5SHGpfBFBM1Sw-ks5qXN4CgaJpZM4HoAcC>
> .
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

mk01 commented:
@fritsch 

indeed there is. would be nice to merge this as well https://github.com/xbmc/xbmc/pull/8549

I updated it by adding #ifdef HAS_IMXVPU to avoid compilation error (imx-cec is not merged upstream).

MartijnKaijser commented:
good for merge? IMX isn't our "official" build so doesn't hurt to merge i guess
jenkins build this please

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/RenderManager.cpp:
```diff
@@ -828,8 +828,15 @@ void CRenderManager::FlipPage(volatile std::atomic_bool& bStop, double pts /* =
         presentmethod = PRESENT_METHOD_BOB;
         invert = true;
       }
+      else if (interlacemethod == VS_INTERLACEMETHOD_IMX_ADVMOTION_DOUBLE)
```
no platform dependent methods here.

mk01 replied:
@FernetMenta 

would it be possible for instance extend AutoInterlaceMethod() with optional parameter - mInt could be passed and platform dependent methods could be mapped to one of that general type?

FernetMenta replied:
the current implementation of AutoInterlaceMethod is maximum nonsense becuase it has nothing to do with an auto-setting. don't touch this.
what you actually want to achive is having RenderManager do a single or double pass depending on deinterlacing method. why don't you ask renderer for the desird method?

mk01 replied:
@FernetMenta 

that was the idea, to reuse silly AutoInterlaceMethod to ask renderer as the call ends there ...
what is the way you have on your mind?

FernetMenta replied:
> that was the idea, to reuse silly AutoInterlaceMethod

don't reuse code that is silly

do something like
bool CBaseRenderer::WantsDoublePass()

that can be overridden by hw renderer

mk01 replied:
@FernetMenta 

reviewed the whole process again - simplified the logic, functionality stays, even removed the last platform specific method from RenderManager's code. 

can you please do a quick check now? the relevant squash is https://github.com/xbmc/xbmc/pull/9253/commits/f450b9e6ae6bc2d0752a2083cdab6379e7c9a4fa

FernetMenta replied:
looks ok from RenderManager's pov
thanks

fritsch commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererIMX.cpp:
```diff
@@ -178,25 +184,25 @@ bool CRendererIMX::RenderUpdateVideoHook(bool clear, DWORD flags, DWORD alpha)
+        // CXBMCRenderManager::PresentFields() is swapping field flag for frame1
+        // this makes IPU render same picture as before, just shifted one line.
+        // let's correct this
+        fieldFmt ^= RENDER_FLAG_FIELDMASK;
```
Is this something one should adress in PresentFields?

mk01 replied:
I find it (personally) odd - and the IPU deint code for BOB too (as it is internally doing the same again - so finally rendering same frame, starting at line 2 instead of 1).

On the other hand the GLES code is handling (this) same output from RenderManager and produced picture is fine. At least on v14 (when IMX had GLES rendering too - I reused AML code for IMX back than - and that code is still used for AML).

... someone would have already realised if similar issue would exists for other implementations (I think).

mk01 committed:
[RenderIMX] remove Supports(EDEINTERLACEMODE mode)

mk01 committed:
[CodecIMX] update fps detection

mk01 committed:
[IMXCodec] add player process info

mk01 committed:
[IMX] fix to lkFIFO

mk01 committed:
[CodecIMX] refactor struct IPUtask * to shared_ptr

mk01 committed:
[CodecIMX] fix logging (LOGVIDEO->LOGDEBUG)

mk01 committed:
[IMX] Interlacing modes update (add advmotion full)

mk01 commented:
@FernetMenta 
as per your idea with the 'wantsdoublepass', would this be ok ?
https://github.com/mk01/xbmc/commit/c57f070a2f640e7baaa178daecd6516da841ad45

FernetMenta commented:
@mk01 yes, without the "invert" parameter. see my comment on the commit

mk01 commented:
@FernetMenta 
ok, thanks.
relevant commits:
https://github.com/xbmc/xbmc/pull/9253/commits/da255d208654fef3c34bff16713d90c5aede7fd6
https://github.com/xbmc/xbmc/pull/9253/commits/26313c869aa3f3bb471da596951f9f27c9b49663

FernetMenta commented:
@mk01 looks good

mk01 committed:
[Renderer] implement WantsDoublePass()

FernetMenta commented:
@mk01 status? is this ready from your side?

mk01 commented:
@FernetMenta 
just added one more commit (updated capture & screenshot function https://github.com/xbmc/xbmc/pull/9253/commits/1e89f920dec6fefe5302e6f812ff2d14a8459ac3).

and yes, it can go in.

FernetMenta commented:
jenkins build this please

fritsch commented:
@mk01 thanks very much! to keep this platform alive.

mhueske commented:
@mk01 Since this PR got merged I notice that playback (especially HD content) is stuttering sometimes as if it skips every second frame. It says it is using deinterlacing mode "BOB", while the same content played on Kodi before this PR it says "hardware" and plays smooth. I am using Cubox-I2ex and LibreELEC with kernel 3.14. Is there something kernel related, as I know you prefer 4.4 kernel? Thanks a lot
